File: sjf.py
def sjf(arrival_time, burst_time, number_of_processes):

  pq = []
  for i in range(number_of_processes):
    pq.append((arrival_time[i], burst_time[i], i))

  completion_times = []
  current_time = 0
  while len(pq) > 0:
    process = pq.pop(0)
    if arrival_time[process[2]] <= current_time:
      current_time += process[1]
      completion_times.append(current_time)
    else:
      current_time = arrival_time[process[2]] + process[1]
      completion_times.append(current_time)
    for i in range(len(pq)):
      if pq[i][1] > process[1]:
        pq[i], pq[0] = pq[0], pq[i]

  turn_around_times = []
  waiting_times = []
  for i in range(number_of_processes):
    turn_around_times.append(completion_times[i] - arrival_time[i])
    waiting_times.append(turn_around_times[i] - burst_time[i])

  average_waiting_time = sum(waiting_times) / number_of_processes
  average_turn_around_time = sum(turn_around_times) / number_of_processes

  return completion_times, turn_around_times, waiting_times, average_waiting_time, average_turn_around_time


if __name__ == "__main__":
  arrival_time = [0, 1, 2, 3]
  burst_time = [4, 5, 2, 1]
  number_of_processes = 4

  completion_times, turn_around_times, waiting_times, average_waiting_time,average_turn_around_time = sjf(arrival_time, burst_time, number_of_processes)
  print("Completion times:", completion_times)
  print("Turn around times:", turn_around_times)
  print("Waiting times:", waiting_times)
  print("Average waiting time:", average_waiting_time)
  print("Average turn around time:", average_turn_around_time)

File: sjf.c
#include <stdio.h>

void swap(int *xp, int *yp) {
    int temp = *xp;
    *xp = *yp;
    *yp = temp;
}

void sortArrivalBurstTime(int n, int arrivalTime[], int burstTime[]) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arrivalTime[j] > arrivalTime[j + 1]) {
                swap(&arrivalTime[j], &arrivalTime[j + 1]);
                swap(&burstTime[j], &burstTime[j + 1]);
            }
        }
    }
}

void calculateTimes(int n, int arrivalTime[], int burstTime[], int completionTime[], int waitingTime[], int turnaroundTime[]) {
    int currentTime = 0;
    int remainingTime[n];
    for (int i = 0; i < n; i++) {
        remainingTime[i] = burstTime[i];
    }

    int completed = 0;

    while (completed < n) {
        int shortestJob = -1;
        int shortestTime = 9999;

        for (int i = 0; i < n; i++) {
            if (arrivalTime[i] <= currentTime && remainingTime[i] < shortestTime && remainingTime[i] > 0) {
                shortestJob = i;
                shortestTime = remainingTime[i];
            }
        }

        if (shortestJob == -1) {
            currentTime++;
        } else {
            remainingTime[shortestJob]--;

            if (remainingTime[shortestJob] == 0) {
                completed++;
                completionTime[shortestJob] = currentTime + 1;
                turnaroundTime[shortestJob] = completionTime[shortestJob] - arrivalTime[shortestJob];
                waitingTime[shortestJob] = turnaroundTime[shortestJob] - burstTime[shortestJob];
            }

            currentTime++;
        }
    }
}

int main() {
    int n;
    printf("Enter the number of processes: ");
    scanf("%d", &n);

    int arrivalTime[n];
    int burstTime[n];
    int completionTime[n];
    int waitingTime[n];
    int turnaroundTime[n];

    printf("Enter arrival time and burst time for each process:\n");
    for (int i = 0; i < n; i++) {
        printf("Process %d:\n", i + 1);
        printf("Arrival Time: ");
        scanf("%d", &arrivalTime[i]);
        printf("Burst Time: ");
        scanf("%d", &burstTime[i]);
    }

    sortArrivalBurstTime(n, arrivalTime, burstTime);
    calculateTimes(n, arrivalTime, burstTime, completionTime, waitingTime, turnaroundTime);

    printf("\nProcess\t Arrival Time\t Burst Time\t Completion Time\t Waiting Time\t Turnaround Time\n");
    for (int i = 0; i < n; i++) {
        printf("%d\t %d\t\t %d\t\t %d\t\t %d\t\t %d\n",
               i + 1, arrivalTime[i], burstTime[i], completionTime[i], waitingTime[i], turnaroundTime[i]);
    }

    // Calculate and display average waiting time and average turnaround time
    float avgWaitingTime = 0, avgTurnaroundTime = 0;
    for (int i = 0; i < n; i++) {
        avgWaitingTime += waitingTime[i];
        avgTurnaroundTime += turnaroundTime[i];
    }
    avgWaitingTime /= n;
    avgTurnaroundTime /= n;

    printf("\nAverage Waiting Time: %.2f\n", avgWaitingTime);
    printf("Average Turnaround Time: %.2f\n", avgTurnaroundTime);

    return 0;
}
